MekHQ/src/mekhq/campaign/Campaign.java

```diff
@@ -3847,6 +3854,57 @@ && getLocation().getJumpPath().getLastSystem().getId().equals(contract.getSystem
         }
     }
 
+    /**
+     * Processes an abandoned convoy. The player is presented with a defeat dialog,
+     * and checks each player template force in the scenario for being a convoy force. If it is a
+     * convoy force, its units are treated as abandoned units.
+     * Each crew member of these units is set as either KIA or POW based on a die roll.
+     * It finally removes each unit from the campaign.
+     *
+     * @param contract The current {@link AtBContract}.
+     * @param scenario The relevant {@link AtBDynamicScenario}.
+     */
+    private void processAbandonedConvoy(AtBContract contract, AtBDynamicScenario scenario) {
+        convoyFinalMessageDialog(this, contract.getEmployerFaction());
+
+        for (Integer forceId : scenario.getPlayerTemplateForceIDs()) {
+            try {
+                Force force = getForce(forceId);
+
+                if (force.isConvoyForce()) {
+                    for (UUID unitID : force.getAllUnits(false)) {
+                        Unit unit = getUnit(unitID);
+
+                        for (Person crewMember : unit.getCrew()) {
+                            decideCrewMemberFate(crewMember);
+                        }
+
+                        removeUnit(unitID);
+                    }
+                }
+            } catch (Exception ex) {
+                logger.warn(ex.getMessage(), ex);
+            }
+        }
+    }
```

Comment:
Wholeheartedly agree. The issue we have right now is that campaign.java is so all encompassing and touching so many different aspects, the likelihood that a new function will be related to something in campaign.java (and therefore logically placed _in_ that class) is fairly high.

Something I think we'd really benefit from is someone ripping campaign.java apart and democratizing a lot of the methods housed there.

Reply:
Campaign became one of two "façade" objects for the project, MHQ being the other one. It is fine, and it happens, however this also means that our efforts should be focused on every time we "touch" the campaing, instead of adding code to it, we take the code that we want to interact with and move to a new place. It is better if campaign become something like a "context", as if we had to communicate with a database and campaign were that database.